### PR TITLE
tp/pp to ep

### PR DESCRIPTION
**Pull request by chai-xiaonan**
1.use load tp/pp ckpt to ep, must add arg --convert-expert-parallel and --no-load-rng,
2.If the --convert-expert-parallel option is enabled, Megatron will execute the modified code; otherwise, it will execute the native Megatron code.
3.For example, after running with the tp2 or pp2 parallel strategy, it is converted to the parallel strategy ep4.

### CODE REVIEW

**Comment by chai-xiaonan** [comment]
收到，谢谢

File: flagscale/backends/Megatron-LM/megatron/training/arguments.py.patch [hunk 1]
```diff
@@ -390,7 +390,16 @@ index e36803fc..3a19a933 100644
      group.add_argument('--exit-on-missing-checkpoint', action='store_true',
                         help="If '--load' is set, but checkpoint is not found "
                         "(e.g., path typo), then exit instead of random "
-@@ -2293,7 +2436,7 @@ def _add_distributed_args(parser):
+@@ -2196,6 +2339,8 @@ def _add_checkpointing_args(parser):
+     group.add_argument('--load-model-opt-format', action='store_true',
+                        help='Load a checkpoint for TensorRT model optimizer (nvidia-modelopt).'
+                             'This function can also be used to load NeMo .nemo sharded checkpoints.')
++    group.add_argument('--convert-expert-parallel', action='store_true', default=None,
++                       help='Loading the ckpt, convert the keys of TP/PP to the keys of EP.')
```
[the remark this answers is not in the document]
Comment (chai-xiaonan) [reply flagscale/backends/Megatron-LM/megatron/training/arguments.py.patch]: The operation is carried out in the way of adding the arg option in megatron. If the action is of bool type, then the action is 'store_true', and this option is not enabled by default

File: flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch [hunk 2]
```diff
@@ -41,7 +49,146 @@ index b5bc894c..93c8b45e 100644
...
++        #    return
++        #if len(self.chained_optimizers) != len(state_dict):
++        #    raise RuntimeError(
++        #        f'Expected {len(self.chained_optimizers)} entries'
++        #        f' in state dict, but got {len(state_dict)}.'
++        #    )
++        #if isinstance(state_dict, dict):
++        #    state_dict = (v for k, v in sorted(state_dict.items()))
++        #for optimizer, state in zip(self.chained_optimizers, state_dict):
++        #    optimizer.load_state_dict(state)
```
[the remark this answers is not in the document]
Comment (chai-xiaonan) [reply flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch]: done

File: flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch [hunk 3]
```diff
@@ -41,7 +49,146 @@ index b5bc894c..93c8b45e 100644
...
++                    sharded_state_dict[optimizer_idx] = optim_state_dict
++                return sharded_state_dict
++            #sharded_state_dict = {}
++            #for optimizer_idx, optimizer in enumerate(self.chained_optimizers):
++            #    optim_state_dict = optimizer.sharded_state_dict(
++            #        model_sharded_state_dict, is_loading, **kwargs
++            #    )
++            #    add_prefix_for_sharding(optim_state_dict, f'chained_{optimizer_idx}.')
++            #    sharded_state_dict[optimizer_idx] = optim_state_dict
++            #return sharded_state_dict
```
[the remark this answers is not in the document]
Comment (chai-xiaonan) [reply flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch]: done

File: flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch [hunk 4]
```diff
@@ -41,7 +49,146 @@ index b5bc894c..93c8b45e 100644
...
++                        model_sharded_state_dict, is_loading, **kwargs
++                    )    
++                    add_prefix_for_sharding(optim_state_dict, f'chained_{optimizer_idx}.')
++                    sharded_state_dict[optimizer_idx] = optim_state_dict
++                return sharded_state_dict
++                 
++            args = get_args()
++            if args.convert_expert_parallel:
++               #convert tp/pp ckpt to ep ckpt
++                print("sharded_state_dict:convert tp/pp ckpt to ep ckpt!")
```
[the remark this answers is not in the document]
Comment (chai-xiaonan) [reply flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch]: done

File: flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch [hunk 5]
```diff
@@ -41,7 +49,146 @@ index b5bc894c..93c8b45e 100644
...
++                        
++                    fake_sharded_state_dict['optimizer'] = optim_state_dict['optimizer']
++                    fake_sharded_state_dict['param_state_sharding_type'] = optim_state_dict['param_state_sharding_type']
++                    
++                    sharded_state_dict[optimizer_idx] = {}
++                    sharded_state_dict[optimizer_idx]['optimizer'] = copy.deepcopy(optim_state_dict['optimizer'])
++                    sharded_state_dict[optimizer_idx]['param_state_sharding_type'] = copy.deepcopy(optim_state_dict['param_state_sharding_type'])
++                    sharded_state_dict[optimizer_idx]['len_param_state'] = len(optim_state_dict['param_state'])
++                    self.original_sharded_state_dict = sharded_state_dict
++                self.mapping_idx = mapping_idx
```
[the remark this answers is not in the document]
Comment (chai-xiaonan) [reply flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch:120]: These two parameters are the key to ensuring accuracy

File: flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch [hunk 6]
```diff
@@ -41,7 +49,146 @@ index b5bc894c..93c8b45e 100644
...
+-            return sharded_state_dict
++            if not is_loading:
++                sharded_state_dict = {} 
++                for optimizer_idx, optimizer in enumerate(self.chained_optimizers):
++                    optim_state_dict = optimizer.sharded_state_dict(
++                        model_sharded_state_dict, is_loading, **kwargs
++                    )    
++                    add_prefix_for_sharding(optim_state_dict, f'chained_{optimizer_idx}.')
++                    sharded_state_dict[optimizer_idx] = optim_state_dict
++                return sharded_state_dict
```
[the remark this answers is not in the document]
Comment (chai-xiaonan) [reply flagscale/backends/Megatron-LM/megatron/core/optimizer/optimizer.py.patch]: This code is designed to ensure the universality of accessing ckpt

**Comment by lxd-cumt** [comment]
Please remove the argument convert-expert-parallel
In load_checkpoint, the parallel strategy used in the state_dict and the one specified in the runtime args can both be obtained within the load_checkpoint function itself; there is no need to introduce an additional argument.

**Comment by lxd-cumt** [comment]
截屏2025-09-18 10 10 18

**Review by lxd-cumt — APPROVED** [review]
LGTM
